Assignment2Q2.java

```diff
@@ -0,0 +1,112 @@
+import java.util.LinkedList;
+
+/**
+ * Created by Alexandra on 23.05.2017.
+ */
+public class Assignment2Q2 {
+
+    public static class TreeNode <T> {
+        private TreeNode<T> left;
+        private TreeNode<T> right;
+        private T value;
+
+        public TreeNode(TreeNode<T> left, TreeNode<T> right, T value) {
+            if (value == null) {
+                throw new IllegalArgumentException();
+            }
+            this.left = left;
+            this.right = right;
+            this.value = value;
+        }
+
+        public TreeNode<T> getLeft() {
+            return left;
+        }
+
+        public TreeNode<T> getRight() {
+            return right;
+        }
+
+        public T getValue() {
+            return value;
+        }
+
+        public String toString() {
+            return value.toString() + '\t';
+        }
+
+        public T lowestCommonAncestor(T value1, T value2) {
+            // since there is only one path from a node to the root
+            // the lowest ancestor will be on this path
+            // and stack of ancestors will also include it
+            LinkedList<TreeNode<T>> stack = new LinkedList<>();
+            LinkedList<Integer> checkedChildren = new LinkedList<>();
+            // 0 - there is no any children equal to value1 or value2
+            // 1 - there is one child equal to value1 or value2
+            // 2 - there are both requested values as children
+            // so we are looking for the first node with mark 2
+            LinkedList<Integer> marks = new LinkedList<>();

```

Comment:
Same comment about having a list of data structure classes, rather than 3 lists, applies.